File: compiler/cbits/genSym.c
#include <Rts.h>
#include <assert.h>
#include "Unique.h"
#include <ghcversion.h>

// These global variables have been moved into the RTS.  It allows them to be
// shared with plugins even if two different instances of the GHC library are
// loaded at the same time (#19940)
//
// The CPP is thus about the RTS version GHC is linked against, and not the
// version of the GHC being built.

#if MIN_VERSION_GLASGOW_HASKELL(9,9,0,0)
// Unique64 patch was present in 9.10 and later
#define HAVE_UNIQUE64 1
#elif !MIN_VERSION_GLASGOW_HASKELL(9,9,0,0) && MIN_VERSION_GLASGOW_HASKELL(9,8,4,0)
// Unique64 patch was backported to 9.8.4
#define HAVE_UNIQUE64 1
#elif !MIN_VERSION_GLASGOW_HASKELL(9,7,0,0) && MIN_VERSION_GLASGOW_HASKELL(9,6,7,0)
// Unique64 patch was backported to 9.6.7
#define HAVE_UNIQUE64 1
#endif

#if !defined(HAVE_UNIQUE64)
HsWord64 ghc_unique_counter64 = 0;
#endif

File: compiler/jsbits/genSym.js
//#OPTIONS: CPP
#include "Unique.h"

// We assume that the unique tag occupies less than 32 bits (should be safe)
#define HIGH_UNIQUE_BITS (32 - UNIQUE_TAG_BITS)
#define HIGH_UNIQUE_MASK ((1 << HIGH_UNIQUE_BITS) - 1)

// The 'ghc_unique_inc' and 'ghc_unique_counter64' are in the native RTS. It allows them to be
// shared with plugins even if two different instances of the GHC library are
// loaded at the same time (#19940)
// However, cross compilers do not support plugins so we have moved these globals back
// into the compiler.
var h$ghc_unique_inc       = h$newByteArray(4);
h$ghc_unique_inc.i3[0]     = 1;
var h$ghc_unique_counter64   = h$newByteArray(8);
h$ghc_unique_counter64.i3[0] = 0;
h$ghc_unique_counter64.i3[1] = 0;

function h$genSym() {
  var rl = h$hs_plusWord64(h$ghc_unique_counter64.i3[1] >>> 0, h$ghc_unique_counter64.i3[0] >>> 0, 0, h$ghc_unique_inc.i3[0] >>> 0);
  h$ret1 = (h$ret1 & HIGH_UNIQUE_MASK) >>> 0;
  // h$ret1 contains the higher part (rh)
  h$ghc_unique_counter64.i3[0] = rl | 0;
  h$ghc_unique_counter64.i3[1] = h$ret1 | 0;
  return rl; // h$ret1 still contains rh
}
